File: add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int reminder = 0;
        ListNode* sumNode = new ListNode((reminder+l1->val+l2->val)%10);
        ListNode* temp = sumNode;
        reminder = (reminder+l1->val%10+l2->val%10)/10;
        l1 = l1->next;l2=l2->next;
        while(l1 && l2)
        {
            temp->next = new ListNode((reminder+l1->val+l2->val)%10);
            reminder = (reminder+l1->val%10+l2->val%10)/10;
            l1 = l1->next;l2=l2->next;
            temp = temp->next;
        }
        
        while(l1)
        {
            temp->next = new ListNode((reminder+l1->val)%10);
            reminder = (reminder+l1->val)/10;
            l1 = l1->next;
            temp = temp->next;
        }
        
        while(l2)
        {
            temp->next = new ListNode((reminder+l2->val)%10);
            reminder = (reminder+l2->val)/10;
            l2 = l2->next;
            temp = temp->next;
        }
        
        if(reminder)temp->next = new ListNode(reminder);
        
        return sumNode;
    }
};

File: January LeetCoding Challenge/Add Two Numbers.java
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return addTwoNumbersWithCarry(l1,l2,0);
    }
    
    public ListNode addTwoNumbersWithCarry(ListNode l1, ListNode l2,int carry) {
        if(l1 == null && l2 == null && carry == 0)return null;
        if(l1 == null)l1 = new ListNode();
        if(l2 == null)l2 = new ListNode();
        int val = l1.val + l2.val + carry;
        ListNode l3 = new ListNode(val%10);
        l3.next = addTwoNumbersWithCarry(l1.next,l2.next,val > 9 ? 1:0);
        return l3;
    }
}
